### PR TITLE
Budget validation and integration tests

### PR DESCRIPTION
**Pull request by danielslbsb**
### Description of changes
* Add integration tests to verify AWS Budget creation and update based on the cluster configuration file.

### Tests
* Integration tests - budget creation, deletion, and update. Uses boto3 to describe budgets and verify changes.

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/templates/budget_builder.py [hunk 1]
```diff
@@ -33,7 +34,9 @@ def _add_parameters(self, budget, index):
             name = f"{self.cluster_config.cluster_name}-{index}-custom-{self.cluster_config.region}"
 
         budget_data = budgets.CfnBudget.BudgetDataProperty(
-            budget_name=name,
+            # Some fields need replacement to be updated iin Cfn. The name without a hash would create duplicates.
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py]: typo inn

File: tests/iam_policies/user-role.cfn.yaml [hunk 2]
```diff
@@ -731,6 +731,11 @@ Resources:
             Resource:
               - !Sub arn:${AWS::Partition}:s3:::integ-tests-*
             Effect: Allow
+          # Required to create and modify budgets
+          - Action:
+              - budgets:*
```
Comment (chenwany) [inline tests/iam_policies/user-role.cfn.yaml]: Do we need all the budget policies? Can we restrict it to only some budget policies?

Comment (chenwany) [reply tests/iam_policies/user-role.cfn.yaml]: The policy also need to be add to the API user stack: https://github.com/aws/aws-parallelcluster/blob/337490be88f21b1517766cd1b45757ebb1c3ad0f/api/infrastructure/parallelcluster-api.yaml#L287

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import boto3
+import pytest
+from assertpy import assert_that
+
+
+@pytest.mark.usefixtures("os", "instance", "scheduler")
+def test_update_add_and_remove_budget(pcluster_config_reader, clusters_factory, test_datadir, region):
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py]: Shall we combine this test with test_update_budget_properties?

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    account_id = sts_client.get_caller_identity().get("Account")
+
+    # Try to describe the budget created by the cluster
+    budget_name_beginning = f"{cluster.name}-0-{region}"
+    budget_client = boto3.client("budgets")
+    budgets_description = budget_client.describe_budgets(AccountId=account_id)["Budgets"]
+    assert_that(
+        any([budget["BudgetName"].startswith(budget_name_beginning) for budget in budgets_description])
+    ).is_true()
+
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py:50]: We can also check the budget start with cluster.name, then check the length of the budget list

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    # Update the cluster to change the budget properties
+    new_config_file = pcluster_config_reader(config_file="all_budget_categories_updated.yaml")
+    cluster.update(str(new_config_file), force_update="true")
+
+    # The budgets from the updated configuration
+    budget_list = cluster.config["DevSettings"]["Budgets"]
+    custom_budget = budget_list[0]
+    cluster_budget = budget_list[1]
+    queue_budget = budget_list[2]
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py]: nit: the code can be combined into one line: 
custom_budget, cluster_budget, queue_budget = budget_list[0], budget_list[1], budget_list[2]

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
+import boto3
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py:12]: Shall we add copyright at the top?

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    sts_client = boto3.client("sts")
+    account_id = sts_client.get_caller_identity().get("Account")
+    budget_client = boto3.client("budgets")
+
+    # Create the initial cluster without a budget definition
+    init_config_file = pcluster_config_reader(config_file="no_budgets.yaml")
+    cluster = clusters_factory(init_config_file)
+
+    # Add a budget, verify that it is added, then delete it and check it was successfully deleted
+    _create_and_delete_budgets(cluster, init_config_file, budget_client, account_id, region, pcluster_config_reader)
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py]: nit: naming: `_test_create_and_delete_budgets`

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    # Add one budget of each BudgetCategory, update them, and check that the fields match the expected values
+    _update_cluster_and_check_fields(cluster, budget_client, account_id, region, pcluster_config_reader)
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py]: nit: naming: `_test_update_cluster_and_check_fields`

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            cluster_budget_name = name
+        elif name.startswith(queue_budget_name_short):
+            queue_budget_name = name
+
+    # Describe the updated budgets to verify that the properties were indeed updated
+    custom_budget_description = budget_client.describe_budget(
+        AccountId=account_id,
+        BudgetName=custom_budget_name,
+    )["Budget"]
+    cluster_budget_description = budget_client.describe_budget(
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py]: ["BudgetLimit"]["Amount"] is already included in `describe_budgets`, we don't need to get name from `describe_budgets` then get limit from `describe_budget`

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    # The expected budget names (without the hash), based on budget_builder.py
+    custom_budget_name_short = f"{cluster.name}-0-custom-{region}"
+    cluster_budget_name_short = f"{cluster.name}-1-{region}"
+    queue_budget_name_short = f"{cluster.name}-2-queue1-{region}"
+
+    # Look for the full names
+    for budget in budget_description:
+        name = budget["BudgetName"]
+        if name.startswith(custom_budget_name_short):
+            custom_budget_name = name
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py:85]: also: 
custom_budget_amount = budget["BudgetLimit"]["Amount"]

File: tests/integration-tests/tests/budgets/test_budgets.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+
+def _check_notification_fields(expected_budget, budget_name, account_id, budget_client):
+    """Check the basic notification fields."""
+    # Each budget in this test's config file has only one notification. Otherwise, comparing them would be difficult
+    budget_notification_description = budget_client.describe_notifications_for_budget(
+        AccountId=account_id,
+        BudgetName=budget_name,
+    )
+    described_notification = budget_notification_description["Notifications"][0]
```
Comment (chenwany) [inline tests/integration-tests/tests/budgets/test_budgets.py]: nit: These can be combined to one line

Comment (danielslbsb) [reply tests/integration-tests/tests/budgets/test_budgets.py]: Done!

File: tests/integration-tests/tests/budgets/test_budgets.py (hunk 10 above)
Comment (danielslbsb) [reply tests/integration-tests/tests/budgets/test_budgets.py:85]: Updated!

File: tests/integration-tests/tests/budgets/test_budgets.py (hunk 9 above)
Comment (danielslbsb) [reply tests/integration-tests/tests/budgets/test_budgets.py]: Good suggestion. Updated!

File: tests/integration-tests/tests/budgets/test_budgets.py (hunk 8 above)
Comment (danielslbsb) [reply tests/integration-tests/tests/budgets/test_budgets.py]: Updated!

File: tests/integration-tests/tests/budgets/test_budgets.py (hunk 6 above)
Comment (danielslbsb) [reply tests/integration-tests/tests/budgets/test_budgets.py:12]: Good reminder! Done.

File: tests/integration-tests/tests/budgets/test_budgets.py (hunk 7 above)
Comment (danielslbsb) [reply tests/integration-tests/tests/budgets/test_budgets.py]: Updated!

**Review by chenwany — APPROVED** [review]

**Pull request merged**